nav2_behavior_tree/plugins/decorator/speed_controller.cpp

```diff
@@ -0,0 +1,107 @@
+// Copyright (c) 2018 Intel Corporation
+// Copyright (c) 2020 Sarthak Mittal
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <string>
+
+#include "nav2_util/geometry_utils.hpp"
+
+#include "speed_controller.hpp"
+
+namespace nav2_behavior_tree
+{
+
+SpeedController::SpeedController(
+  const std::string & name,
+  const BT::NodeConfiguration & conf)
+: BT::DecoratorNode(name, conf),
+  first_time_(false),
+  period_(1.0)
+{
+  getInput("min_rate", min_rate_);
+  getInput("max_rate", max_rate_);
+  getInput("min_speed", min_speed_);
```

Comment:
Wouldn't the minimum just be 0? That does ask the question: what if this is negative for an extended duration?

Reply:
I added the minimum speed param as a way to clip the rate below a certain traveling speed. For example, if min_speed was set to 0.5 m/s, the rate will be clamped to min_rate at any speed <=0.5

Reply:
Got it, good idea.

Reply:
I think you should expose these ports in the BT when we use it. I want people to specifically see what the min/max speed/rates are. If a robot goes any normal speed (e.g. non-turtlebot) 0.25 is nothing. Maybe bump that default max to 0.5 m/s.